### DISCUSSION TITLE
BIN ENGINES

### DISCUSSION THREAD

**Discussion by Pepebotika1980** [eh2k/squares-and-circles#111] [General]
What is the BIN engine and what files are loaded? thank you

**Comment by eh2k** [comment]
hi @Pepebotika1980,

squeres and circles has an internal 'file system' for extras, which in principle a table that can only be expanded. When you flash/upload a BIN file, the first 8 characters of the file name are used as a key and added to the table. The data is written in the free memory area. The firmware itself is located at the beginning, the ‘file system’ is at the end of the flash memory and is expanded backwards until at some point it theoretically overwrites the firmware.

There is a total of approx. 1 MB free memory (Teensy 4.0) that can be written in this way at the moment .
You cannot delete individual blobs, only the entire table. (Format SD command)

In this way the DX7 sysex or sample ROMs for the TR707 or TR909 Highhats are stored, that are not part of the firmware.
The bin-blobs can be accessed in this way: https://github.com/eh2k/squares-and-circles/blob/main/app/DRUMS/TR707.cpp#L23

**Reply by Pepebotika1980** [reply]
Thanks for the reply, I really don't understand anything from the link you sent me, I see that they are programming issues that I don't understand. Updating the firmware, loading the Fv1, 707, 909 and Sysex files is already under control, I just needed to know if that section of the BIN allowed me to load more things and maybe I was wasting it. But I'm not going to touch things that I don't control. Thank you very much for the great work you do, it's an illusion every time I see that there is an update. Thanks!

**Reply by eh2k** [reply]
I think you're the first asking. Actually, only I used it when developing. Well, here a second explanation for non-programmers. Bin files are actually any files that can be copied to the device (actually ‘FILE’ would be more appropriate than ‘BIN’). If you update the firmware, the files remain. The files can also be executable engines at some point. If I run out of memory or if I no longer want certain engines installed by default, I can simply leave them out. But if someone still wants to use them, they can install them this way.
